manager/orchestrator/replicatedjob/orchestrator.go

```diff
@@ -45,6 +52,162 @@ func (o *Orchestrator) Run(ctx context.Context) {
+				runningTasks++
+				slots[task.Slot] = true
+			}
+		}
+	}
+
+	// now that we have our counts, we need to see how many new tasks to
+	// create. this number can never exceed MaxConcurrent, but also should not
+	// result in us exceeding TotalCompletions. first, get these numbers out of
+	// the service spec.
+	rj := service.Spec.GetReplicatedJob()
+
+	// possibleNewTasks gives us the upper bound for how many tasks we'll
+	// create. also, ugh, subtracting uints. there's no way this can ever go
+	// wrong.
+	possibleNewTasks := rj.MaxConcurrent - runningTasks
+
+	// allowedNewTasks is how many tasks we could create, if there were no
+	// restriction on maximum concurrency. This is the total number of tasks
+	// we want completed, minus the tasks that are already completed, minus
+	// the tasks that are in progress.
+	//
+	// seriously, ugh, subtracting unsigned ints. totally a fine and not at all
+	// risky operation, with no possibility for catastrophe
+	allowedNewTasks := rj.TotalCompletions - completeTasks - runningTasks
+
+	// the lower number of allowedNewTasks and possibleNewTasks is how many we
+	// can create. we'll just use an if statement instead of some fancy floor
+	// function.
+	actualNewTasks := allowedNewTasks
+	if possibleNewTasks < allowedNewTasks {
+		actualNewTasks = possibleNewTasks
+	}
+
+	// this check might seem odd, but it protects us from an underflow of the
+	// above subtractions, which, again, is a totally impossible thing that can
+	// never happen, ever, obviously.
+	if actualNewTasks > rj.TotalCompletions {
+		return fmt.Errorf(
+			"uint64 underflow, we're not going to create %v tasks",
+			actualNewTasks,
+		)
+	}
+
+	// finally, we can create these tasks. do this in a batch operation, to
+	// avoid exceeding transaction size limits
+	err := o.store.Batch(func(batch *store.Batch) error {
+		for i := uint64(0); i < actualNewTasks; i++ {
+			if err := batch.Update(func(tx store.Tx) error {
+				// find an unoccupied slot number that we can use, same as with
+				// replicated jobs. all of the
+				var slot uint64
+				// the total number of slots we can have for a job is equal to
+				// the value of MaxConcurrent for the job iteration. This means
+				// if we iterate over values from 0 to MaxConcurrent, we'll
+				// find an available slot.
+				for s := uint64(0); s < rj.MaxConcurrent; s++ {
+					// when we're iterating through, if the service has slots
+					// that haven't been used yet (for example, if this is the
+					// first time we're running this iteration), then doing
+					// a map lookup for the number will return the 0-value
+					// (false) even if the number doesn't exist in the map.
+					if !slots[s] {
+						slot = s
+						// once we've found a slot, mark it as occupied, so we
+						// don't double assign in subsequent iterations.
+						slots[slot] = true
+						break
+					}
+				}
+
+				task := orchestrator.NewTask(o.cluster, service, slot, "")
```

Comment:
Is there any possibilities for tasks to be created with a certain `nodeId`? Would it make sense to actually set this parameter here with something that's passed from top down and propagated?

I am very familiar with the overall requirements, so this is just one question I had when I saw the empty string being passed in.

Reply:
In this case, no. The Node ID is only passed in for global services and will be passed in for global jobs. In replicated services and jobs, the Node ID is blank. I would need to check, but I'm pretty sure the global service in turn passes only the Node ID and not the slot.